vars/pythonSetupPyPipeline.groovy

```diff
@@ -49,96 +61,212 @@ def call(Map pipelineParams) {
         }
-      }
+      } // Bump version Release
 
       stage("Build") {
-        steps {
-          sh "python setup.py build"
-        }
-      }
-
-      stage("Execute Pylint") {
-        steps {
-          // Necessary to avoid issues with creating pylint files that track the delta of warnings.
-          withEnv(["HOME=$WORKSPACE"]) {
-            // setuptools-lint does not make a good job in installing all its dependencies.
-            sh "python setup.py install --user"
-            sh "python setup.py lint --lint-output-format parseable"
+        agent {
+          dockerfile {
+            filename pipelineParams.dockerFilename
+            additionalBuildArgs pipelineParams.dockerBuildArgs
+            args pipelineParams.dockerRunArgs
+            reuseNode true
           }
         }
-        post {
-          always {
-            recordIssues enabledForFailure: true, tool: pyLint(), healthy: 0
-          }
-        }
-      }
+        stages {
+          stage("Build") {
+            parallel {
+              stage("Build module") {
+                steps {
+                  sh "python setup.py build"
 
-      stage("Execute Tests") {
-        steps {
-          sh "python setup.py test --addopts '--cov-report xml:build/coverage.xml --cov-report term --cov-branch --junitxml=build/test_results.xml'"
+                  script {
+                    // Gather the module name and version so that it can be used in a later stage
+                    moduleName = sh(script: "python setup.py --name", returnStdout: true).trim()
+                    moduleVersion = sh(script: "python setup.py --version", returnStdout: true).trim()
+                    echo("Module name: ${moduleName} version: ${moduleVersion}")
+                  }
+                }
+              } // Build module
+
+              stage("Build docs") {
+                steps {
+                  // Necessary to be able to locally install packages, which is required for build_sphinx
+                  withEnv(["HOME=$WORKSPACE"]) {
+                    sh "python setup.py install --user"
+                    sh "python setup.py build_sphinx"
+                  }
+                }
+              } // Build docs
+            }
+          } // Build
         }
-        post {
-          always {
-            junit 'build/test_results.xml'
-            cobertura coberturaReportFile: 'build/coverage.xml'
+      } // Build
+
+      stage("Validation") {
+        agent {
+          dockerfile {
+            filename pipelineParams.dockerFilename
+            additionalBuildArgs pipelineParams.dockerBuildArgs
+            args pipelineParams.dockerRunArgs
+            reuseNode true
           }
         }
-      }
+        stages {
+          stage("Validate") {
+            parallel {
+              stage("Run Linter") {
+                steps {
+                  // Necessary to avoid issues with creating pylint files that track the delta of warnings.
```

Comment:
we should have clean builds (with a clean workspace) for each pipeline run

Reply:
That is handled on BitBucket team project level already.

That step is just needed to make pylint happy (being able to create a file)

Reply:
Wiping the workspace and docker images in the post cleanup stage